txpool/txpool.go

```diff
@@ -724,90 +739,111 @@ func (p *TxPool) pruneAccountsWithNonceHoles() {
 // (only once) and an enqueueRequest is signaled.
 func (p *TxPool) addTx(origin txOrigin, tx *types.Transaction) error {
 	if p.logger.IsDebug() {
-		p.logger.Debug("add tx",
-			"origin", origin.String(),
-			"hash", tx.Hash.String(),
-		)
+		p.logger.Debug("add tx", "origin", origin.String(), "hash", tx.Hash.String())
 	}
 
 	// validate incoming tx
 	if err := p.validateTx(tx); err != nil {
 		return err
 	}
 
+	// calculate tx hash
+	tx.ComputeHash()
+
+	// initialize account for this address once or retrieve existing one
+	account, _ := p.createAccountOnce(tx.From)
+	// populate currently free slots
+	slotsFree := p.gauge.max - p.gauge.read()
+
+	account.promoted.lock(true)
+	account.enqueued.lock(true)
+	account.nonceToTx.lock()
+
+	defer func() {
+		account.nonceToTx.unlock()
+		account.enqueued.unlock()
+		account.promoted.unlock()
+	}()
+
+	accountNonce := account.getNonce()
+
+	//	only accept transactions with expected nonce
 	if p.gauge.highPressure() {
 		p.signalPruning()
 
-		//	only accept transactions with expected nonce
-		if account := p.accounts.get(tx.From); account != nil &&
-			tx.Nonce > account.getNonce() {
+		if tx.Nonce > accountNonce {
 			metrics.IncrCounter([]string{txPoolMetrics, "rejected_future_tx"}, 1)
 
 			return ErrRejectFutureTx
 		}
 	}
 
+	// try to see if there is transaction with same nonce for this account
+	oldTxWithSameNonce := account.nonceToTx.get(tx.Nonce)
+	if oldTxWithSameNonce != nil {
+		if oldTxWithSameNonce.Hash == tx.Hash {
+			metrics.IncrCounter([]string{txPoolMetrics, "already_known_tx"}, 1)
+
+			return ErrAlreadyKnown
+		} else if oldTxWithSameNonce.GasPrice.Cmp(tx.GasPrice) >= 0 {
+			// if tx with same nonce does exist and has same or better gas price -> return error
+			return ErrUnderpriced
+		}
+
+		slotsFree += slotsRequired(oldTxWithSameNonce) // add old tx slots
```

Comment:
If we get to this point, means the new transaction is accepted and will replace previous one, then in line 806 we check `slotsFree` but containing old a new tx slots added, this is not correct because the new slots will substitute old slots, if I'm not wrong.

Reply:
we must increment available slots by number of slots occupied by old tx. thats why this line exists

Reply:
we still dont know at this point that new tx will be accepted because number of slots it used can be too high for example

Reply:
Ah yes you're right I misread the line, we're reserving the space.